lib/Dialect/SPIRV/Transforms/DecorateSPIRVCompositeTypeLayoutPass.cpp

```diff
@@ -0,0 +1,318 @@
+
+#include "mlir/Dialect/SPIRV/Passes.h"
+#include "mlir/Dialect/SPIRV/SPIRVDialect.h"
+#include "mlir/Dialect/SPIRV/SPIRVOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+/// According to the Vulkan spec "14.5.4. Offset and Stride Assignment":
+/// "There are different alignment requirements depending on the specific
+/// resources and on the features enabled on the device."
+///
+/// There are 3 types of alignment: scalar, base, extended.
+/// See the spec for details.
+///
+/// Note: Even if scalar alignment is supported, it is generally more
+/// performant to use the base alignment. So here the calculation is based on
+/// base alignment.
+///
+/// The memory layout must obey the following rules:
+/// 1. The Offset decoration of any member must be a multiple of its alignment.
+/// 2. Any ArrayStride or MatrixStride decoration must be a multiple of the
+/// alignment of the array or matrix as defined above.
+/// 3. The ArrayStride, MatrixStride, and Offset decorations must be large
+/// enough to hold the size of the objects they affect (that is, specifying
+/// overlap is invalid).
+class VulkanLayoutUtils {
+public:
+  using Alignment = uint64_t;
+
+  /// Returns a new type with layout info. Assigns the type size in bytes to the
+  /// `size`. Assigns the type alignment in bytes to the `alignment`.
+  static Type decorateType(spirv::StructType structType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  /// Checks whether a type is legal in terms of Vulkan layout info
+  /// decoration. A type is dynamically legal if it's a composite type in the
+  /// StorageBuffer, PhysicalStorageBuffer, Uniform, and PushConstant Storage
+  /// Classes without layout informtation.
+  static bool isLegalType(Type type);
+
+private:
+  static Type decorateType(Type type, spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  static Type decorateType(VectorType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  static Type decorateType(spirv::ArrayType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  /// Calculates the alignment for the given scalar type.
+  static Alignment getScalarTypeAlignment(Type scalarType);
+};
+
+Type VulkanLayoutUtils::decorateType(
+    spirv::StructType structType, spirv::StructType::LayoutInfo &size,
+    VulkanLayoutUtils::Alignment &alignment) {
+  if (!structType.getNumElements()) {
+    return structType;
+  }
+
+  llvm::SmallVector<Type, 4> memberTypes;
+  llvm::SmallVector<spirv::StructType::LayoutInfo, 4> layoutInfo;
+  llvm::SmallVector<spirv::StructType::MemberDecorationInfo, 4>
+      memberDecorations;
+
+  spirv::StructType::LayoutInfo structMemberOffset = 0;
+  spirv::StructType::LayoutInfo memberSize = 0;
```

Comment:
Define `memberSize` and `memberAlignment` inside the for loop so we don't need to reset them at the end?